application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramContentHandler.java

```diff
@@ -0,0 +1,209 @@
+        XWikiDocument oldDoc = context.getWiki().getDocument(oldDocRef, context);
+        XWikiDocument newDoc = context.getWiki().getDocument(newDocRef, context);
+
+        BufferedReader reader = new BufferedReader(new InputStreamReader(attachment.getContentInputStream(context)));
+        StringBuffer sb = new StringBuffer();
+
+        String str;
+        while ((str = reader.readLine()) != null) {
+            sb.append(str);
+        }
+
+        String regex = "\"%s\"";
+        String oldAbsoluteURL = String.format(regex, oldDoc.getExternalURL(VIEW_ACTION, context));
+        String newAbsoluteURL = String.format(regex, newDoc.getExternalURL(VIEW_ACTION, context));
+        String oldURL = String.format(regex, oldDoc.getURL(VIEW_ACTION, context));
+        String newURL = String.format(regex, newDoc.getURL(VIEW_ACTION, context));
+
+        String newContent = sb.toString().replaceAll(oldAbsoluteURL, newAbsoluteURL).replaceAll(oldURL, newURL);
+
+        attachment.setContent(new ByteArrayInputStream(newContent.getBytes()));
+
+        return attachment;
+    }
+
+    /**
+     * Update content of the diagram with new links.
+     * 
+     * @param backlinkDoc document that has a backlink to the diagram
+     * @param originalDocRef reference of the document before rename
+     * @param currentDocRef reference of the document after rename
+     * @param context context of the execution
+     * @throws ParserConfigurationException if a DocumentBuilder cannot be created
+     * @throws SAXException if parsing the document fails
+     * @throws IOException if any IO errors occur
+     * @throws XWikiException if saving the document fails
+     */
+    public void updateDiagramContent(XWikiDocument backlinkDoc, DocumentReference originalDocRef,
+        DocumentReference currentDocRef, XWikiContext context)
+        throws ParserConfigurationException, SAXException, IOException, XWikiException
+    {
+        Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder()
+            .parse(new ByteArrayInputStream(backlinkDoc.getContent().getBytes()));
+
+        NodeList userObjectList = document.getElementsByTagName("UserObject");
+        for (int i = 0; i < userObjectList.getLength(); i++) {
+            linkHandler.updateUserObjectNode(userObjectList.item(i), currentDocRef, originalDocRef);
+        }
+
+        NodeList mxCellList = document.getElementsByTagName("mxCell");
+        for (int i = 0; i < mxCellList.getLength(); i++) {
+            linkHandler.updateMxCellNode(mxCellList.item(i), currentDocRef, originalDocRef);
+        }
+
+        backlinkDoc.setContent(getStringFromDocument(document));
+        context.getWiki().saveDocument(backlinkDoc, "Updated diagram after page rename", context);
+    }
+
+    /**
+     * Convert a document to String.
+     * 
+     * @param doc the Document to be processed
+     * @return converted Document to String
+     */
+    public String getStringFromDocument(Document doc)
```

Comment:
You should look into [XMLUtils.serialize()](https://github.com/xwiki/xwiki-commons/blob/master/xwiki-commons-core/xwiki-commons-xml/src/main/java/org/xwiki/xml/XMLUtils.java). Could be used for parsing the XML also.